src/middlewared/middlewared/plugins/activedirectory.py

```diff
@@ -0,0 +1,1147 @@
+        self.ad = kwargs.get('ad_conf')
+        self.hosts = kwargs.get('hosts')
+        self.interfaces = kwargs.get('interfaces')
+        self.logger = kwargs.get('logger')
+        self.pagesize = 1024
+        self._isopen = False
+        self._handle = None
+        self._rootDSE = None
+        self._rootDomainNamingContext = None
+        self._configurationNamingContext = None
+        self._defaultNamingContext = None
+        
+        return
+
+    def __enter__(self):
+        return self
+    
+    def __exit__(self, typ, value, traceback):
+        if typ is not None:
+            raise
+
+    def validate_credentials(self):
+        """
+        For credential validation we simply open an ldap connection
+        """
+        ret = self._open()
+        if ret:
+            self._close()
+        return ret
+
+    def _open(self):
+        """
+        We can only intialize a single host. In this case,
+        we iterate through a list of hosts until we get one that
+        works and then use that to set our LDAP handle.
+
+        SASL GSSAPI bind only succeeds when DNS reverse lookup zone
+        is correctly populated. Fall through to simple bind if this
+        fails.
+        """
+        res = None 
+        if self._isopen:
+            return True
+
+        if self.hosts:
+            saved_simple_error = None
+            saved_gssapi_error = None
+            for server in self.hosts:
+                proto = 'ldaps' if SSL(self.ad['ssl']) == SSL.USESSL else 'ldap' 
+                uri = f"{proto}://{server['host']}:{server['port']}"
+                try:
+                    self._handle = ldap.initialize(uri)
+                except Exception as e:
+                    self.logger.debug(f'Failed to initialize ldap connection to [{uri}]. Moving to next server.')
+                    continue
+
+                if self.ad['verbose_logging']:
+                    self.logger.debug(f'Successfully initialized LDAP server: [{uri}]')
+
+                res = None
+                ldap.protocol_version = ldap.VERSION3
+                ldap.set_option(ldap.OPT_REFERRALS, 0)
+                ldap.set_option(ldap.OPT_NETWORK_TIMEOUT, 10.0)
+
+                if SSL(self.ad['ssl']) != SSL.NOSSL:
+                    ldap.set_option(ldap.OPT_X_TLS_ALLOW, 1)
+                    if self.certfile:
+                        ldap.set_option(
+                            ldap.OPT_X_TLS_CACERTFILE,
+                            self.certfile
+                        )
+                    ldap.set_option(
+                        ldap.OPT_X_TLS_REQUIRE_CERT,
+                        ldap.OPT_X_TLS_ALLOW
+                    )
+
+                if SSL(self.ad['ssl']) == SSL.USESSL:
+                    try:
+                        self._handle.start_tls_s()
+
+                    except ldap.LDAPError as e:
+                        raise CallError(e)
+                        continue
```

Comment:
If we are raising the continue will be useless here

Reply:
Yeah, not sure what I was thinking here. I've decided to come down on one side of the fence instead of trying to lie down on it.